src/examples/zkapps/dex/dex.ts

```diff
@@ -19,250 +19,339 @@ import {
+  class Dex extends SmartContract {
+    // addresses of token contracts are constants
+    tokenX = addresses.tokenX;
+    tokenY = addresses.tokenY;
+
+    /**
+     * state which keeps track of total lqXY supply -- this is needed to calculate what to return when redeeming liquidity
+     *
+     * total supply is zero initially; it increases when supplying liquidity and decreases when redeeming it
+     */
+    @state(UInt64) totalSupply = State<UInt64>();
+
+    /**
+     * Mint liquidity tokens in exchange for X and Y tokens
+     * @param user caller address
+     * @param dx input amount of X tokens
+     * @param dy input amount of Y tokens
+     * @return output amount of lqXY tokens
+     *
+     * This function fails if the X and Y token amounts don't match the current X/Y ratio in the pool.
+     * This can also be used if the pool is empty. In that case, there is no check on X/Y;
+     * instead, the input X and Y amounts determine the initial ratio.
+     *
+     * The transaction needs to be signed by the user's private key.
+     */
+    @method supplyLiquidityBase(
+      user: PublicKey,
+      dx: UInt64,
+      dy: UInt64
+    ): UInt64 {
+      let tokenX = new TokenContract(this.tokenX);
+      let tokenY = new TokenContract(this.tokenY);
+
+      // get balances of X and Y token
+      // TODO: this creates extra account updates. we need to reuse these by passing them to or returning them from transfer()
+      // but for that, we need the @method argument generalization
+      let dexXBalance = tokenX.getBalance(this.address);
+      let dexYBalance = tokenY.getBalance(this.address);
+
+      // // assert dy == [dx * y/x], or x == 0
+      let isXZero = dexXBalance.equals(UInt64.zero);
+      let xSafe = Circuit.if(isXZero, UInt64.one, dexXBalance);
+
+      // FIXME
+      // Error: Constraint unsatisfied (unreduced): Equal 0 1
+      dy.equals(dx.mul(dexYBalance).div(xSafe)).or(isXZero).assertTrue();
+
+      tokenX.transfer(user, this.address, dx);
+      tokenY.transfer(user, this.address, dy);
+
+      // calculate liquidity token output simply as dl = dx + dx
+      // => maintains ratio x/l, y/l
+      let dl = dy.add(dx);
+      let userUpdate = this.experimental.token.mint({
+        address: user,
+        amount: dl,
+      });
+      if (lockedLiquiditySlots !== undefined) {
+        /**
+         * exercise the "timing" (vesting) feature to lock the received liquidity tokens.
+         *
+         * THIS IS HERE FOR TESTING!
+         *
+         * In reality, the timing feature is a bit awkward to use for time-locking liquidity tokens.
+         * That's because, if there is currently a vesting schedule on an account, we can't modify it.
+         * Thus, a liquidity provider would need to wait for their current tokens to unlock before being able to
+         * supply liquidity again (or, create another account to supply liquidity from).
+         */
+        let amountLocked = dl;
```

Comment:
Should we spin up a ticket to add a nicer API for updating timing functionality?

Reply:
It's covered by this one: https://github.com/o1-labs/snarkyjs/issues/186